### PR TITLE
refactor(editor): Unify ConnectionType and NodeConnectionTypes (no-changelog)

### PR DESCRIPTION
**Pull request by RicardoE105**
## Summary

I was moving the `NodeSettings.vue` file [here](https://github.com/n8n-io/n8n/pull/10545) to use the composition API and trying to fix one small type [issue](https://github.com/n8n-io/n8n/actions/runs/10583349351/job/29325171428#step:8:115) snowball into a lot of type issues because we have two types in the FE that represent the same. One is an enum, and the other is a type union that we use in the editor-ui, workflow, and nodes-base, and they do not match (a known problem with enums). So, to avoid this, simplify the types and update the code case to use it.

Types changed from:

```js
export type ConnectionTypes =
	| 'ai_agent'
	| 'ai_chain'
	| 'ai_document'
	| 'ai_embedding'
	| 'ai_languageModel'
	| 'ai_memory'
	| 'ai_outputParser'
	| 'ai_retriever'
	| 'ai_textSplitter'
	| 'ai_tool'
	| 'ai_vectorRetriever'
	| 'ai_vectorStore'
	| 'main';

export const enum NodeConnectionType {
	AiAgent = 'ai_agent',

	AiChain = 'ai_chain',

	AiDocument = 'ai_document',

	AiEmbedding = 'ai_embedding',

	AiLanguageModel = 'ai_languageModel',

	AiMemory = 'ai_memory',

	AiOutputParser = 'ai_outputParser',

	AiRetriever = 'ai_retriever',

	AiTextSplitter = 'ai_textSplitter',

	AiTool = 'ai_tool',

	AiVectorStore = 'ai_vectorStore',

	Main = 'main',
}

export const nodeConnectionTypes: NodeConnectionType[] = [
	NodeConnectionType.AiAgent,
	NodeConnectionType.AiChain,
	NodeConnectionType.AiDocument,
	NodeConnectionType.AiEmbedding,
	NodeConnectionType.AiLanguageModel,
	NodeConnectionType.AiMemory,
	NodeConnectionType.AiOutputParser,
	NodeConnectionType.AiRetriever,
	NodeConnectionType.AiTextSplitter,
	NodeConnectionType.AiTool,
	NodeConnectionType.AiVectorStore,
	NodeConnectionType.Main,
];
```

to:

```js
export const NodeConnectionType = {
	AiAgent: 'ai_agent',
	AiChain: 'ai_chain',
	AiDocument: 'ai_document',
	AiEmbedding: 'ai_embedding',
	AiLanguageModel: 'ai_languageModel',
	AiMemory: 'ai_memory',
	AiOutputParser: 'ai_outputParser',
	AiRetriever: 'ai_retriever',
	AiTextSplitter: 'ai_textSplitter',
	AiTool: 'ai_tool',
	AiVectorStore: 'ai_vectorStore',
	AiVectorRetriever: 'ai_vectorRetriever',
	Main: 'main',
} as const;

export type NodeConnectionTypes = (typeof NodeConnectionType)[keyof typeof NodeConnectionType];
```

It seems having enums is quite discouraged in TS:

- https://fettblog.eu/tidy-typescript-avoid-enums/
- https://www.executeprogram.com/blog/typescript-features-to-avoid
- https://dev.to/ylerjen/use-typescript-const-assertion-instead-of-enums-mfn

### Update:

Doing this comes with the tradeoff of not being able to use the `NodeConnectionType` in expressions within the node properties. That means we need to set the connection types directly as strings. Considering this rarely changes and there are only a few instances where we use it in the `nodes-packages` and `nodes-langchain`, I still think this is better than having two different types in the `workflow` and `editor-ui` packages that represent the same (enum NodeConnectionType and type ConnectionTypes) but are incompatible, causing a bunch of type issues.

Removed all instances of the new `NodeConnectionType` from `nodes-base` and `nodes-langchain`, even though it works everywhere but in expressions, to prevent people from misusing it in the expressions. It's still type-checked.

Wondering if this shifts the problem from one package to another instead of improving the situation as a whole. Maybe instead of adding the `const NodeConectionType` object only use the `enum NodeConnectionType` everywhere, and get rid of the `ConnectionType` only.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

**Comment by RicardoE105** [comment]
@alexgrozav, it looks like somehow I broke the Webhook node connection, but I have not been able to pinpoint the issue. It must be something small that I'm overlooking. If you have any pointers, they will be highly appreciated 🙏 

image

**Review by alexgrozav — CHANGES_REQUESTED** [review]
Listed a few things that I think we should fix.

File: packages/editor-ui/src/components/RunDataAi/useAiContentParsers.ts [hunk 1]
```diff
@@ -213,9 +227,11 @@ export type ParsedAiContent = Array<{
 export const useAiContentParsers = () => {
 	const parseAiRunData = (
 		executionData: INodeExecutionData[],
-		endpointType: NodeConnectionType,
+		endpointType: NodeConnectionTypes,
 	): ParsedAiContent => {
-		if ([NodeConnectionType.AiChain, NodeConnectionType.Main].includes(endpointType)) {
+		if (
+			([NodeConnectionType.AiChain, NodeConnectionType.Main] as string[]).includes(endpointType)
```
[the remark this answers is not in the document]
Comment (RicardoE105) [reply packages/editor-ui/src/components/RunDataAi/useAiContentParsers.ts:222]: Without it 
image

**Comment by RicardoE105** [comment]
@OlegIvaniv figured out the issue: Changing from enum to const caused the issue. We were using the enum in the expression in the nodes-base package, specifically in some of the node properties. Within the expressions, the enums work because they are resolved at build time. When using const the expressions do not have access to the reference (the object) because they run in a sandbox.

See conversation below:

> Oleg -> We're using NodeConnectionType in several places as const enum to get the appropriate connection key, but it's no longer exported
-> Ricardo -> but I updated it all the places with the new type
-> Oleg -> Yeah but you've changed it from const enum
-> Ricardo -> but it's the same you can do NodeConnectionType.AiAgent, and it s the same with the enum or with the object, both are going to return the string value.
-> Oleg -> Not in the expressions, though; those are not compiled. The as const would preserve the reference of an object during run-time while const enum would replace it inline during build. And when resolving expressions we don't have access to that reference
-> Oleg -> And some inputs/outputs are expressions
-> Oleg -> So when they're resolved, it looks like this:
You can see that there's a reference to n8n_workflow and window but again, those are not available in the sandbox when resolving expressions


image

**Comment by RicardoE105** [comment]
closing in favor of https://github.com/n8n-io/n8n/pull/10595/files

**Comment by RicardoE105** [comment]
closing in favor of https://github.com/n8n-io/n8n/pull/10595

**Pull request closed without merging**